File: CLI/2019/Test/FK_CLI_Shadow/Shadow.cs
using System;
using FK_CLI;

namespace FK_CLI_Shadow
{
    class Shadow
    {
        const int WIN_W = 1024;
        const int WIN_H = 1024;

        static void ModelSetup(fk_Model argModel, fk_Material argMat, fk_Vector argPos)
        {
            argModel.Material = argMat;
            argModel.SmoothMode = true;
            argModel.GlMoveTo(argPos);
        }

        static void BallMove(fk_AppWindow argWindow, fk_Model argModel)
        {
            const double STEP = 0.5;

            if (argWindow.GetSpecialKeyStatus(fk_Key.UP, fk_Switch.PRESS))
            {
                argModel.GlTranslate(0.0, 0.0, -STEP);
            }
            if (argWindow.GetSpecialKeyStatus(fk_Key.DOWN, fk_Switch.PRESS))
            {
                argModel.GlTranslate(0.0, 0.0, STEP);
            }
            if (argWindow.GetSpecialKeyStatus(fk_Key.RIGHT, fk_Switch.PRESS))
            {
                argModel.GlTranslate(STEP, 0.0, 0.0);
            }
            if (argWindow.GetSpecialKeyStatus(fk_Key.LEFT, fk_Switch.PRESS))
            {
                argModel.GlTranslate(-STEP, 0.0, 0.0);
            }
        }

        static void RobotRotate(fk_AppWindow argwindow, fk_Model argModel)
        {
            var origin = new fk_Vector(0.0, 0.0, 0.0);

            if (argwindow.GetKeyStatus('z', fk_Switch.PRESS))
            {
                argModel.LoRotateWithVec(origin, fk_Axis.Y, Math.PI / 60.0);
            }
            if (argwindow.GetKeyStatus('x', fk_Switch.PRESS))
            {
                argModel.LoRotateWithVec(origin, fk_Axis.Y, -Math.PI / 60.0);
            }
        }


        static void Main(string[] args)
        {
            var window = new fk_AppWindow();
            var bgColor = new fk_Color(0.5, 0.5, 0.5);

            fk_Material.InitDefault();
            window.Size = new fk_Dimension(WIN_W, WIN_H);
            window.TrackBallMode = true;
            window.ShowGuide();
            window.CameraPos = new fk_Vector(0.0, 50.0, 200.0);
            window.CameraFocus = new fk_Vector(0.0, 0.0, 0.0);
            window.DefaultLightVec = new fk_Vector(1.0, -1.0, 1.0);
            window.DefaultLightMaterial = fk_Material.WhiteLight;
            window.BGColor = bgColor;
            window.FPS = 60;

            var ifsShape = new fk_IFSTexture();
            if (ifsShape.ReadBMP("00tex_master.BMP") == false)
            {
                Console.WriteLine("Image Read Error");
            }

            if (ifsShape.ReadMQOFile("meka.mqo", "body01") == false)
            {
                Console.WriteLine("ifs load err");
            }

            var sph = new fk_Sphere(8, 7.0);
            var floor = new fk_Block(200.0, 2.0, 200.0);
            var spModel = new fk_Model();
            var ifsModel = new fk_Model();
            var floorModel = new fk_Model();

            spModel.Shape = sph;
            ifsModel.Shape = ifsShape;
            floorModel.Shape = floor;

            spModel.ShadowEffect = true;
            spModel.ShadowDraw = true;
            ifsModel.ShadowEffect = true;
            ifsModel.ShadowDraw = true;
            floorModel.ShadowDraw = true;

            spModel.FogMode = false;

            // 各モデルをディスプレイリストに登録
            window.Entry(floorModel);
            window.Entry(spModel);
            window.Entry(ifsModel);

            //window.ShadowMode = fk_ShadowMode.OFF;
            //window.ShadowMode = fk_ShadowMode.HARD;
            //window.ShadowMode = fk_ShadowMode.SOFT_FAST;
            window.ShadowMode = fk_ShadowMode.SOFT_NICE;
            window.ShadowVec = new fk_Vector(1.0, -1.0, 1.0);
            window.ShadowAreaSize = 500.0;
            window.ShadowDistance = 300.0;
            window.ShadowResolution = 1024;
            window.ShadowVisibility = 0.7;
            window.ShadowBias = 0.01;

            window.FogMode = fk_FogMode.EXP2;
            window.FogDensity = 0.005;
            window.FogColor = bgColor;

            ModelSetup(spModel, fk_Material.Yellow, new fk_Vector(-20.0, 20.0, 0.0));
            ModelSetup(ifsModel, fk_Material.White, new fk_Vector(20.0, 5.0, 0.0));
            ModelSetup(floorModel, fk_Material.White, new fk_Vector(0.0, -1.0, 0.0));


            window.Open();
            for(int count = 0;  window.Update(); count++)
            {
                BallMove(window, spModel);
                RobotRotate(window, ifsModel);
                if (count == 300) ifsModel.ShadowEffect = false;
            }
        }
    }
}

File: build_vs/2019/test/FK_Test_Shadow/shadow.cxx
#include <FK/FK.h>

using namespace std;
using namespace FK;

const int WIN_W = 1024; // ウィンドウ横幅
const int WIN_H = 1024; // ウィンドウ縦幅

// 通常モデルのシェーダー設定
void ModelSetup(fk_Model *argModel, fk_Material argMat, fk_Vector argPos)
{
	argModel->setMaterial(argMat);
	argModel->setSmoothMode(true);
	argModel->glMoveTo(argPos);
}

void BallMove(fk_AppWindow *argWindow, fk_Model *argModel)
{
	double step = 0.5;

	if(argWindow->getSpecialKeyStatus(fk_Key::UP, fk_Switch::PRESS)) {
		argModel->glTranslate(0.0, 0.0, -step);
	}
	if(argWindow->getSpecialKeyStatus(fk_Key::DOWN, fk_Switch::PRESS)) {
		argModel->glTranslate(0.0, 0.0, step);
	}
	if(argWindow->getSpecialKeyStatus(fk_Key::RIGHT, fk_Switch::PRESS)) {
		argModel->glTranslate(step, 0.0, 0.0);
	}
	if(argWindow->getSpecialKeyStatus(fk_Key::LEFT, fk_Switch::PRESS)) {
		argModel->glTranslate(-step, 0.0, 0.0);
	}
}

void RobotRotate(fk_AppWindow *argWindow, fk_Model *argModel)
{
	fk_Vector origin;
	if(argWindow->getKeyStatus('z', fk_Switch::PRESS)) {
		argModel->loRotateWithVec(origin, fk_Axis::Y, fk_Math::PI/60.0);
	}
	if(argWindow->getKeyStatus('x', fk_Switch::PRESS)) {
		argModel->loRotateWithVec(origin, fk_Axis::Y, -fk_Math::PI/60.0);
	}
}

void ShadowChange(int argCount, fk_Model *argModel)
{
	if((argCount/100) % 2 == 0) {
		argModel->setShadowEffect(true);
	} else {
		argModel->setShadowEffect(false);
	}
}

int main(int, char **)
{
	fk_System::setcwd();
	Error::SetMode(Error::Mode::BROWSER_INTERACTIVE);

	fk_AppWindow *window = new fk_AppWindow();
	fk_Material *mat = new fk_Material();

	fk_IFSTexture *ifsShape = new fk_IFSTexture();
	fk_Sphere *sph = new fk_Sphere(8, 7.0);
	fk_Block *floor = new fk_Block(200.0, 2.0, 200.0);
	fk_Model *spModel = new fk_Model();
	fk_Model *ifsModel = new fk_Model();
	fk_Model *floorModel = new fk_Model();

	fk_Color bgColor(0.5, 0.5, 0.5);

	fk_Material::initDefault();
	window->setSize(WIN_W, WIN_H);
	window->setTrackBallMode(true);
	window->showGuide();

	window->setCameraPos(0.0, 50.0, 200.0);
	window->setCameraFocus(0.0, 0.0, 0.0);
	window->setDefaultLightVec(1.0, -1.0, 1.0);
	window->setDefaultLightMaterial(Material::WhiteLight);

	window->setFogMode(fk_FogMode::EXP2);
	window->setFogColor(bgColor);
	window->setFogLinearMap(20.0, 300.0);
	window->setFogDensity(0.005);

	// モデル
	if(ifsShape->readBMP("data/model/00tex_master.BMP") == false) {
		fl_alert("tex load err");
	}

	if(ifsShape->readMQOFile("data/model/meka.mqo", "body01") == false) {
		fl_alert("ifs load err");
	}

	ifsShape->setTexRendMode(fk_TexRendMode::SMOOTH);
	spModel->setShape(sph);
	ifsModel->setShape(ifsShape);
	floorModel->setShape(floor);

	spModel->setShadowEffect(true);
	spModel->setShadowDraw(true);
	ifsModel->setShadowEffect(false);
	ifsModel->setShadowDraw(true);
	floorModel->setShadowDraw(true);

	//spModel->setFogMode(false);
	ifsModel->setFogMode(false);
	
	// 各モデルをディスプレイリストに登録
	window->setBGColor(bgColor);
	window->entry(floorModel);
	window->entry(spModel);
	window->entry(ifsModel);

	//fk_ShadowMode mode = fk_ShadowMode::OFF;
	//fk_ShadowMode mode = fk_ShadowMode::HARD;
	//fk_ShadowMode mode = fk_ShadowMode::SOFT_FAST;
	fk_ShadowMode mode = fk_ShadowMode::SOFT_NICE;
	window->setShadowMode(mode);
	window->setShadowVec(1.0, -1.0, 1.0);
	window->setShadowAreaSize(500.0);
	window->setShadowDistance(300.0);
	window->setShadowResolution(1024);
	window->setShadowVisibility(1.0);
	window->setShadowBias(0.005);

	ModelSetup(spModel, Material::Yellow, fk_Vector(-20.0, 20.0, 0.0));
	ModelSetup(ifsModel, Material::White, fk_Vector(20.0, 5.0, 0.0));
	ModelSetup(floorModel, Material::White, fk_Vector(0.0, -1.0, 0.0));

	window->open();
	for(int count = 0; window->update(); ++count) {
		BallMove(window, spModel);
		RobotRotate(window, ifsModel);
		ShadowChange(count, ifsModel);
	}


	delete floorModel;
	delete ifsModel;
	delete spModel;
	delete floor;
	delete sph;
	delete ifsShape;
	delete mat;
	delete window;

	return 0;
}
